lib/auth/keystore/gcp_kms.go

```diff
@@ -0,0 +1,454 @@
+	req := &kmspb.GetPublicKeyRequest{
+		Name: keyID.keyVersionName,
+	}
+	// Retry fetching the public key while the key is pending creation.
+	resp, err := retryWhilePending(ctx, g.clock, g.kmsClient.GetPublicKey, req)
+	if err != nil {
+		return nil, trace.Wrap(err, "unexpected error fetching public key")
+	}
+
+	block, _ := pem.Decode([]byte(resp.Pem))
+	if block == nil {
+		return nil, trace.BadParameter("GCP KMS key %s has invalid public key PEM", keyID.keyVersionName)
+	}
+	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return nil, trace.Wrap(err, "unexpected error parsing public key pem")
+	}
+
+	return &kmsSigner{
+		kmsClient: g.kmsClient,
+		clock:     g.clock,
+		keyID:     keyID,
+		public:    pub,
+	}, nil
+}
+
+func (s *kmsSigner) Public() crypto.PublicKey {
+	return s.public
+}
+
+func (s *kmsSigner) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	ctx := context.Background()
+
+	requestDigest := &kmspb.Digest{}
+	switch opts.HashFunc() {
+	case crypto.SHA256:
+		requestDigest.Digest = &kmspb.Digest_Sha256{
+			Sha256: digest,
+		}
+	case crypto.SHA512:
+		requestDigest.Digest = &kmspb.Digest_Sha512{
+			Sha512: digest,
+		}
+	default:
+		return nil, trace.BadParameter("unsupported hash func for GCP KMS signer: %v", opts.HashFunc())
+	}
+
+	resp, err := doGCPRequest(ctx, s.clock, s.kmsClient.AsymmetricSign, &kmspb.AsymmetricSignRequest{
+		Name:   s.keyID.keyVersionName,
+		Digest: requestDigest,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "error while attempting GCP KMS signing operation")
+	}
+	return resp.Signature, nil
+}
+
+type gcpKMSKeyID struct {
+	keyVersionName string
+}
+
+func (g gcpKMSKeyID) marshal() []byte {
+	var buf bytes.Buffer
+	buf.WriteString(gcpkmsPrefix)
+	buf.WriteString(g.keyVersionName)
+	return buf.Bytes()
+}
+
+func parseGCPKMSKeyID(key []byte) (gcpKMSKeyID, error) {
+	var keyID gcpKMSKeyID
+	if keyType(key) != types.PrivateKeyType_GCP_KMS {
+		return keyID, trace.BadParameter("unable to parse invalid GCP KMS key")
+	}
+	// strip gcpkms: prefix
+	keyID.keyVersionName = string(key[len(gcpkmsPrefix):])
+	return keyID, nil
+}
+
+// withTimeout is like context.WithTimeout, but with mockable time
+func withTimeout(ctx context.Context, clock clockwork.Clock, d time.Duration) (context.Context, func()) {
+	ctx, ctxCancel := context.WithCancel(ctx)
+	timer := clock.NewTimer(d)
+	cancel := func() {
+		ctxCancel()
+		timer.Stop()
+	}
+	go func() {
+		select {
+		case <-timer.Chan():
+		case <-ctx.Done():
+		}
+		cancel()
+	}()
+	return ctx, cancel
+}
+
+func retryWhilePending[reqType, optType, respType any](
+	ctx context.Context,
+	clock clockwork.Clock,
+	f func(context.Context, reqType, ...optType) (*respType, error),
+	req reqType,
+) (*respType, error) {
+	ctx, cancel := withTimeout(ctx, clock, defaultGCPPendingTimeout)
+	defer cancel()
+
+	ticker := clock.NewTicker(defaultGCPPendingRetryInterval)
+	defer ticker.Stop()
+	for {
```

Comment:
Would it make sense to cap the amount of retries to some reasonably huge number here?

Reply:
It's effectively capped by `defaultGCPPendingTimeout / defaultGCPPendingRetryInterval`, which is currently 12.